### PR TITLE
(CDAP-20) Remove almost all guava usages from cdap-api

### PR DESCRIPTION
**Pull request by chtyim**
- Almost all usages are remove except the following:
  - TypeToken
  - BiMap and related implementation
  - Multimap and related implementation

- For TypeToken usages, it's all internal now (meaning not leaking through any interfaces), hence we can just copy the TypeToken (and related classes) into cdap-api
- For BiMap usages, we can replace it with two maps or copy it from guava as well
- For Multimap, it is problematic since it is leaked as part of public API through HttpServiceRequest/Responder. We can do deprecation, however it means we cannot remove guava dependency until two versions later, which we needed for the Cassandra usage case now.

### CODE REVIEW

File: cdap-api/src/main/java/co/cask/cdap/api/data/schema/Schema.java [hunk 1]
```diff
@@ -775,15 +839,14 @@ private String buildString() {
...
-    try {
-      SCHEMA_TYPE_ADAPTER.write(writer, this);
-      writer.close();
-      return builder.toString();
+
+    StringWriter writer = new StringWriter();
+    try (JsonWriter jsonWriter = new JsonWriter(writer)) {
+      SCHEMA_TYPE_ADAPTER.write(jsonWriter, this);
     } catch (IOException e) {
       // It should never throw IOException on the StringBuilder Writer, if it does, something very wrong.
```
Comment (gokulavasan) [inline cdap-api/src/main/java/co/cask/cdap/api/data/schema/Schema.java]: comment update: StringWriter instead of StringBuilderWriter

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/KeyValueTable.java [hunk 2]
```diff
@@ -226,7 +225,9 @@ public void write(KeyValue<byte[], byte[]> keyValue) throws IOException {
       private boolean closed = false;
       @Override
       protected KeyValue<byte[], byte[]> computeNext() {
-        Preconditions.checkState(!closed);
+        if (closed) {
```
Comment (gokulavasan) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/KeyValueTable.java]: Why the change in logic?

Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/KeyValueTable.java]: The old logic is not good. Should always use the endOfData to signal end of data instead of throwing illegal state exception.

**Comment by gokulavasan** [comment]
LGTM :+1:

File: cdap-api/src/main/java/co/cask/cdap/api/data/schema/Schema.java (hunk 1 above)
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/data/schema/Schema.java]: fixed

**Comment by chtyim** [comment]
Build passed: http://builds.cask.co/browse/CDAP-DUT2451-7
